examples/exex/rollup/src/db.rs

```diff
@@ -0,0 +1,482 @@
+    )?;
+
+    Ok(())
+}
+
+/// Delete account by address. Connection can be either [rusqlite::Transaction] or
+/// [rusqlite::Connection].
+fn delete_account<C: Deref<Target = Connection>>(
+    connection: &C,
+    address: Address,
+) -> eyre::Result<()> {
+    connection.execute("DELETE FROM account WHERE address = ?", (address.to_string(),))?;
+    Ok(())
+}
+
+/// Get account by address using the database connection. Connection can be either
+/// [rusqlite::Transaction] or [rusqlite::Connection].
+fn get_account<C: Deref<Target = Connection>>(
+    connection: &C,
+    address: Address,
+) -> eyre::Result<Option<AccountInfo>> {
+    match connection.deref().query_row::<String, _, _>(
+        "SELECT data FROM account WHERE address = ?",
+        (address.to_string(),),
+        |row| row.get(0),
+    ) {
+        Ok(account_info) => Ok(Some(serde_json::from_str(&account_info)?)),
+        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
+        Err(e) => Err(e.into()),
+    }
+}
+
+/// Insert new storage if it does not exist, update otherwise. Connection can be either
+/// [rusqlite::Transaction] or [rusqlite::Connection].
+fn upsert_storage<C: Deref<Target = Connection>>(
+    connection: &C,
+    address: Address,
+    key: B256,
+    data: U256,
+) -> eyre::Result<()> {
+    connection.execute(
+        "INSERT INTO storage (address, key, data) VALUES (?, ?, ?) ON CONFLICT(address, key) DO UPDATE SET data = excluded.data",
+        (address.to_string(), key.to_string(), data.to_string()),
+    )?;
+    Ok(())
+}
+
+/// Delete storage by address and key. Connection can be either [rusqlite::Transaction] or
+/// [rusqlite::Connection].
+fn delete_storage<C: Deref<Target = Connection>>(
+    connection: &C,
+    address: Address,
+    key: B256,
+) -> eyre::Result<()> {
+    connection.execute(
+        "DELETE FROM storage WHERE address = ? AND key = ?",
+        (address.to_string(), key.to_string()),
+    )?;
+    Ok(())
+}
+
+/// Get all storages for the provided address using the database connection. Connection can be
+/// either [rusqlite::Transaction] or [rusqlite::Connection].
+fn get_storages<C: Deref<Target = Connection>>(
+    connection: &C,
+    address: Address,
+) -> eyre::Result<Vec<(B256, U256)>> {
+    connection
+        .prepare("SELECT key, data FROM storage WHERE address = ?")?
+        .query((address.to_string(),))?
+        .mapped(|row| {
+            Ok((
+                B256::from_str(row.get_ref(0)?.as_str()?),
+                U256::from_str(row.get_ref(1)?.as_str()?),
+            ))
+        })
+        .map(|result| {
+            let (key, data) = result?;
+            Ok((key?, data?))
+        })
+        .collect()
+}
+
+/// Get storage for the provided address by key using the database connection. Connection can be
+/// either [rusqlite::Transaction] or [rusqlite::Connection].
+fn get_storage<C: Deref<Target = Connection>>(
+    connection: &C,
```

Comment:
why not just `Connection`?

Reply:
want to make it with both `Transaction` and `Connection`

Reply:
hmm, since all of them are `Deref` providing the conn as `&` or `&*` should work, so we don't duplicate the functions

Reply:
sometimes we want to call the method directly on `Database` like in the tests https://github.com/paradigmxyz/reth/blob/af521cd3774f94e8f9bd76b2fe1c9aa74c3b38cd/examples/exex/rollup/src/main.rs#L502-L504 and sometimes we use it inside another complex method like here https://github.com/paradigmxyz/reth/blob/af521cd3774f94e8f9bd76b2fe1c9aa74c3b38cd/examples/exex/rollup/src/db.rs#L316-L332

there's no duplication of the function bodies, just the signatures for convenience of calling using a `Database` method instead of a freestanding function